### PR TITLE
Add --resolve to target-query

### PR DESCRIPTION
**Pull request by Miauwkeru**
- Moves all the record modifier code to helpers/modifier.py
- Adds modifier logic, to allow for more modifiers to get added later (currently only paths get used)

- open the target_path during hashing:

When opened in a DirLoader, the OS will "open" the file
However, the file never got closed, so on linux you'd get
an error that the maximum open files threshold was reached

- Add tests for resolve

### CODE REVIEW

File: dissect/target/filesystem.py [hunk 1]
```diff
@@ -834,8 +834,8 @@ def hash(self, algos: Optional[Union[List[str], List[Callable]]] = None) -> tupl
             The various digests of this entry.
         """
         if algos:
-            return hashutil.custom(self.open(), algos)
-        return hashutil.common(self.open())
+            return hashutil.custom(fd, algos)
+        return hashutil.common(fd)
```
Comment (Schamper) [inline dissect/target/filesystem.py:838]: Where is `fd` coming from?

File: dissect/target/filesystem.py
Comment (Schamper) [inline dissect/target/filesystem.py:1]: I can't comment on it, but could you maybe also replace all instances of `List` in this file with `list`?

File: dissect/target/plugins/os/windows/regf/firewall.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def firewall(self):
                     data[fname] = value
 
                 if "app" in data:
-                    data["app"] = self.target.resolve(data["app"])
+                    data["app"] = str(self.target.resolve(data["app"]))
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/firewall.py:70]: Shouldn't the `app` field in the record become a `path`?

File: dissect/target/tools/query.py [hunk 3]
```diff
@@ -349,29 +350,39 @@ def main():
...
-                    else:
-                        raise e
-
-                if break_out:
-                    break
+
+        modifier = None
+
+        if args.resolve:
+            modifier = modifier.Modifier.RESOLVE
```
Comment (Schamper) [inline dissect/target/tools/query.py:357]: Doesn't this conflict with the import name?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/regf/firewall.py [hunk 4]
```diff
@@ -11,7 +11,7 @@ class FirewallPlugin(Plugin):
     """Plugin that parses firewall rules from the registry."""
 
     KEY = "HKLM\\SYSTEM\\CurrentControlSet\\Services\\SharedAccess\\Parameters\\FirewallPolicy\\FirewallRules"
-    FIELD_MAP = {"app": "uri"}
+    FIELD_MAP = {"app": "path"}
```
Comment (pyrco) [inline dissect/target/plugins/os/windows/regf/firewall.py:14]: Are you sure this is a path and not an uri?

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/target/helpers/modifier.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        if path is None:
+            continue
+
+        yield target.resolve(str(path)), field_name
+
+
+def _create_modified_record(record_name: str, field_name: str, type_info: Iterator[tuple[str, str, str]]):
+    record_kwargs = dict()
+    record_def = list()
+    for type, name, field in type_info:
```
Comment (pyrco) [inline dissect/target/helpers/modifier.py:33]:
```suggestion
    for type, name, data in type_info:
```

File: dissect/target/helpers/modifier.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        hashed_field_name = f"{field_name}{name}"
```
Comment (pyrco) [inline dissect/target/helpers/modifier.py:34]:
```suggestion
        extended_field_name = f"{field_name}{name}"
```

File: dissect/target/helpers/modifier.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+def _resolve_path_types(target: Target, record: Record) -> Iterator[tuple[TargetPath, str]]:
+    for field_name, field_type in record._field_types.items():
+        if not issubclass(field_type, fieldtypes.path):
+            continue
+
+        path = getattr(record, field_name, None)
+        if path is None:
+            continue
+
+        yield target.resolve(str(path)), field_name
```
Comment (pyrco) [inline dissect/target/helpers/modifier.py:27]:
```suggestion
        yield field_name, target.resolve(str(path))
```

This order feels more logical (it is more in line with key, value pairs).

File: dissect/target/helpers/modifier.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        else:
+            additional_record.append(_record)
+
+    if not additional_record:
+        return record
+
+    return GroupedRecord(record._desc.name, [record] + additional_record)
+
+
+def _resolve_path_records(resolved_path: TargetPath, field_name: str) -> Record:
```
Comment (pyrco) [inline dissect/target/helpers/modifier.py:78]:
```suggestion
def _resolve_path_records(field_name: str, resolved_path: TargetPath) -> Record:
```
This order feels more logical (it is more in line with key, value pairs).

File: dissect/target/helpers/modifier.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    """Resolve files from path fields inside the record."""
+    type_info = [("path", "_resolved", resolved_path)]
+    return _create_modified_record("filesystem/file/resolved", field_name, type_info)
+
+
+def _hash_path_records(resolved_path: TargetPath, field_name: str) -> Record:
```
Comment (pyrco) [inline dissect/target/helpers/modifier.py:84]:
```suggestion
def _hash_path_records( field_name: str, resolved_path: TargetPath) -> Record:
```
This order feels more logical (it is more in line with key, value pairs).

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/target/helpers/record_modifier.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+        path = getattr(record, field_name, None)
+        if path is None:
+            continue
+
+        yield field_name, target.resolve(str(path))
+
+
+def modify_record(target: Target, record: Record, modifier_function: ModifierFunc) -> GroupedRecord:
+    additional_record = []
```
Comment (pyrco) [inline dissect/target/helpers/record_modifier.py:74]:
```suggestion
    additional_records = []
```

File: dissect/target/helpers/record_modifier.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    return _create_modified_record("filesystem/file/digest", field_name, type_info)
+
+
+MODIFIER_MAPPING = {
+    Modifier.RESOLVE: _resolve_path_records,
+    Modifier.HASH: _hash_path_records,
+}
+
+
+def _resolve_path_types(target: Target, record: Record) -> Iterable[tuple[TargetPath, str]]:
```
Comment (pyrco) [inline dissect/target/helpers/record_modifier.py:61]:
```suggestion
def _resolve_path_types(target: Target, record: Record) -> Iterator[tuple[str, TargetPath]]:
```

This actually is an Iterator ;)

File: dissect/target/helpers/record_modifier.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+ModifierFunc = Callable[[Target, Record], GroupedRecord]
+
+
+class Modifier(StrEnum):
+    RESOLVE = "resolve"
+    HASH = "hash"
+
+
+def _create_modified_record(record_name: str, field_name: str, field_info: Iterable[tuple[str, str, str]]):
```
Comment (pyrco) [inline dissect/target/helpers/record_modifier.py:26]:
```suggestion
def _create_modified_record(record_name: str, field_name: str, field_info: Iterable[tuple[str, str, TargetPath]]):
```

File: dissect/target/helpers/record_modifier.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    record_kwargs = dict()
+    record_def = list()
+    for type, name, data in field_info:
```
Comment (pyrco) [inline dissect/target/helpers/record_modifier.py:29]:
```suggestion
    for type, name_suffix, data in field_info:
```

File: dissect/target/helpers/hashutil.py [hunk 14]
```diff
@@ -58,70 +54,34 @@ def custom(fh: BinaryIO, algos: list[Union[str, HASH]]) -> tuple[str]:
...
 
-def hash_uri_records(target: Target, record: Record) -> Record:
-    """Hash uri paths inside the record."""
+def hash_uri(target: Target, path: str) -> tuple[str, str]:
+    """Hash the target path."""
     warnings.warn(
         (
-            "The hash_uri_records() function is deprecated, and will be removed in dissect.target 3.15. "
-            "Use hash_path_records() instead"
+            "The hash_uri() function is deprecated, and will be removed in dissect.target 3.15."
```
Comment (pyrco) [inline dissect/target/helpers/hashutil.py:61]:
```suggestion
            "The hash_uri() function is deprecated, and will be removed in dissect.target 3.15. "
```

The space should have been there as it's a continuation.

File: tests/helpers/test_modifier.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    record: Record,
+    hash_function: ModifierFunc,
+    target_win: Mock,
+    test_input: dict[str, Union[type[path], type[str]]],
+    expected_records: int,
+) -> None:
+    record._desc.name = "test"
+    record._field_types = test_input
+
+    record_names = [key for key, value in test_input.items() if value is path]
```
Comment (pyrco) [inline tests/helpers/test_modifier.py:48]:
```suggestion
    path_field_names = [key for key, value in test_input.items() if value is path]
```

File: tests/helpers/test_modifier.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        patch.object(TargetPath, "open", mock_open(read_data=b"")),
+        patch("dissect.target.helpers.record_modifier.common", return_value=HASHES),
+    ):
+        hashed_record = hash_function(target_win, record)
+
+    assert hashed_record.name == "test"
+    assert len(hashed_record.records) == expected_records
+    assert hashed_record.records[0] == record
+
+    _record = hashed_record.records[1]
```
Comment (pyrco) [inline tests/helpers/test_modifier.py:60]:
```suggestion
    _record = hashed_record.records[1]
```

This looks like it is not being used.

**Review by pyrco — CHANGES_REQUESTED** [review]

**Review by pyrco — CHANGES_REQUESTED** [review]

**Review by pyrco — APPROVED** [review]

**Pull request merged**